tailnet/conn.go

```diff
@@ -55,6 +56,22 @@ func init() {
 	// Globally disable network namespacing. All networking happens in
 	// userspace.
 	netns.SetEnabled(false)
+	// Tailscale, by default, "trims" the set of peers down to ones that we are "actively" communicating with in
+	// an effort to save memory.  But, we want to make sure the Wireguard connection is up and handshaked before sending
+	// TCP traffic over it to avoid anomalously long round-trip time of the initial handshake
+	// c.f. https://github.com/coder/coder/issues/7388#issuecomment-1625463069 for more details.
+	//
+	// If Tailscale is waiting for traffic to bring up Wireguard, and we wait for Wireguard to send traffic, that's a
+	// deadlock.  So, disable this feature.
+	//
+	// Note that Tailscale.com's use case is very different from ours: in their use case, users create one persistent
+	// tailnet per device, and it allows connections to every other thing in Tailscale that belongs to them.  The
+	// tailnet stays up as long as your laptop or phone is turned on.
+	//
+	// Our use case is different: for clients, it's a point-to-point connection to a single workspace, and lasts only as
+	// long as the connection.  For agents, it's connections to a small number of clients (CLI or Coderd) that are being
+	// actively used by the end user.
+	envknob.Setenv("TS_DEBUG_TRIM_WIREGUARD", "false")
```

Comment:
One thing that comes to mind here is that agents will forever spin trying to connect to CLI connections that have gone away. Since we don't have the ability to tell agents that a client is gone, it seems like this would pile up.

Reply:
Also, is this something we should keep disabled on coderd's tailnet as well? I guess it's sort-of moot since we're doing our own peer expiration.

Reply:
> One thing that comes to mind here is that agents will forever spin trying to connect to CLI connections that have gone away. Since we don't have the ability to tell agents that a client is gone, it seems like this would pile up.

Yeah, they already do that https://github.com/coder/coder/issues/7960

We need to enhance the coordinator to revoke nodes when clients disconnect.

Reply:
> Also, is this something we should keep disabled on coderd's tailnet as well? I guess it's sort-of moot since we're doing our own peer expiration.

I don't think we want to trim wireguard on Coderd's tailnet, since that means we cannot wait for wireguard to come up before attempting to connect to the agent, and we risk hitting stalls due to an inflated retry timeout #7388 .  As you mention, we have to solve the problem of when to remove agent peers in Coderd.  Trimming the wireguard connections saves memory, but we'd still be leaking memory if we just never removed them from tailnet.